sdk/src/main/java/io/customer/sdk/api/HttpRequestRunner.kt

```diff
@@ -0,0 +1,117 @@
+package io.customer.sdk.api
+
+import io.customer.base.extenstions.add
+import io.customer.base.extenstions.hasPassed
+import io.customer.sdk.error.CustomerIOApiErrorResponse
+import io.customer.sdk.error.CustomerIOApiErrorsResponse
+import io.customer.sdk.error.CustomerIOError
+import io.customer.sdk.repository.PreferenceRepository
+import io.customer.sdk.util.JsonAdapter
+import io.customer.sdk.util.Logger
+import io.customer.sdk.util.SimpleTimer
+import retrofit2.Response
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+interface HttpRequestRunner {
+    suspend fun <R> performAndProcessRequest(makeRequest: suspend () -> Response<R>): Result<R>
+}
+
+/**
+ * Where HTTP response processing occurs.
+ */
+class HttpRequestRunnerImpl(
+    private val prefsRepository: PreferenceRepository,
+    private val logger: Logger,
+    private val retryPolicy: HttpRetryPolicy,
+    private val retryPolicyTimer: SimpleTimer,
+    private val jsonAdapter: JsonAdapter
+) : HttpRequestRunner {
+
+    override suspend fun <R> performAndProcessRequest(makeRequest: suspend () -> Response<R>): Result<R> {
+        prefsRepository.httpRequestsPauseEnds?.let { httpPauseEnds ->
+            if (!httpPauseEnds.hasPassed()) {
+                logger.debug("HTTP request ignored because requests are still paused.")
+                return Result.failure(CustomerIOError.HttpRequestsPaused())
+            }
+        }
+
+        val response = makeRequest()
+
+        val responseBody = response.body()
+        if (response.isSuccessful && responseBody != null) {
+            prepareForNextRequest() // after successful HTTP request, reset to prepare for the next HTTP request
+
+            return Result.success(responseBody)
+        }
+
+        return processUnsuccessfulResponse(response, makeRequest)
+    }
+
+    suspend fun <R> processUnsuccessfulResponse(response: Response<R>, makeRequest: suspend () -> Response<R>): Result<R> {
+        when (val statusCode = response.code()) {
+            in 500 until 600 -> {
+                val sleepTime = retryPolicy.nextSleepTime
+                return if (sleepTime != null) {
+                    logger.debug("Encountered $statusCode HTTP response. Sleeping $sleepTime seconds and then retrying.")
+
+                    retryPolicyTimer.scheduleAndCancelPreviousSuspend(sleepTime) {
+                        this.performAndProcessRequest(makeRequest)
+                    }
+                } else {
+                    pauseHttpRequests()
+                    prepareForNextRequest() // after retry policy is finished, reset to prepare for the next HTTP request
+
+                    Result.failure(CustomerIOError.ServerDown())
+                }
+            }
+            401 -> {
+                pauseHttpRequests()
```

Comment:
Why would we `pause` the requests if it's `401` unauthenticated user? 

My use case is if a client has entered wrong workspace credentials, initially they wouldn't get to know about it until we have hit `X seconds` after that we will pause the `http` requests for another `Y` seconds. 

Its going to take atleast X+Y seconds to find out if the requests are going successfully now?

Reply:
This has been discussed outside of GitHub. We decided to move forward with this implementation but create a ticket to see if we can improve it in the future.